autopush/logging.py

```diff
@@ -1,120 +1,130 @@
+    "why",
+])
+
+
+@implementer(ILogObserver)
+class PushLogger(object):
+    def __init__(self, logger_name, log_level="debug", log_format="json",
+                 log_output="stdout", sentry_dsn=None):
+        self.logger_name = "-".join([
+            logger_name,
+            pkg_resources.get_distribution("autopush").version
+        ])
+        self._filename = None
+        self.log_level = LogLevel.lookupByName(log_level)
+        if log_output == "stdout":
+            self._output = sys.stdout
+        else:
+            self._filename = log_output
+            self._output = "file"
+        if log_format == "json":
+            self.format_event = self.json_format
+        else:
+            self.format_event = formatEventAsClassicLogText
+        if sentry_dsn:
             self.raven_client = raven.Client(
                 release=raven.fetch_package_version("autopush"))
         else:
             self.raven_client = None
-        self.logger = Logger()
-
-    def raven_log(self, event):
-        """Log out twisted exception failures to Raven"""
-        f = event['failure']
-        # Throw back to event loop
-        reactor.callFromThread(
-            self.raven_client.captureException,
-            (f.type, f.value, f.getTracebackObject())
-        )
 
-    def __call__(self, msg):
-        """Called to log out messages"""
-        error = bool(msg.get("isError"))
+    def __call__(self, event):
+        if event["log_level"] < self.log_level:
+            return
 
-        if self.raven_client and 'failure' in msg:
-            self.raven_log(msg)
-            error = True
+        if self.raven_client and 'failure' in event:
+            f = event["failure"]
+            reactor.callFromThread(
+                self.raven_client.captureException,
+                (f.type, f.value, f.getTracebackObject())
+            )
 
-        # Twisted log messages on Python 2 are bytes. We don't know the
-        # encoding, but assume it's ASCII superset. Charmap will translate
-        # ASCII correctly, and higher-bit characters just map to
-        # corresponding Unicode code points, and will never fail at decoding.
-        message = unicode(textFromEventDict(msg), "charmap")
-        kw = msg.copy()
-        for key in ["message", "isError", "failure", "why", "format"]:
-            kw.pop(key, None)
-        TWISTED_LOG_MESSAGE(error=error, message=message, **kw).write(
-            self.logger)
+        text = self.format_event(event)
+        self._output.write(unicode(text))
+        self._output.flush()
 
-    def start(self):
-        """Start capturing Twisted logs."""
-        startLoggingWithObserver(self, setStdout=False)
+    def json_format(self, event):
+        error = bool(event.get("isError")) or "failure" in event
+        ts = event["log_time"]
 
-
-def stdout(message):
-    """Format a message appropriately for structured logging capture of stdout
-    and then write it to stdout"""
-    if HUMAN:
```

Comment:
I take it that the new logging format automatically produces messages that are easily scan-able for folk that are debugging?